src/util/strconv.h

```diff
@@ -106,25 +110,98 @@ static char *rm_strndup_unescape(const char *s, size_t len) {
   return ret;
 }
 
+// transform utf8 string to lowercase using nunicode library
+// encoded: the utf8 string to transform, if the transformation is successful
+//          the transformed string will be written back to this buffer
+// in_len: the length of the utf8 string
+// returns the bytes written to encoded, or 0 if the length of the transformed
+// string is greater than in_len and no transformation was done
+static size_t unicode_tolower(char *encoded, size_t in_len) {
+  uint32_t u_stack_buffer[SSO_MAX_LENGTH];
+  uint32_t *u_buffer = u_stack_buffer;
+
+  if (in_len == 0) {
+    return 0;
+  }
+
+  const char *encoded_char = encoded;
+  ssize_t u_len = nu_strtransformnlen(encoded, in_len, nu_utf8_read,
+                                              nu_tolower, nu_casemap_read);
+
+  if (u_len >= (SSO_MAX_LENGTH - 1)) {
+    u_buffer = (uint32_t *)rm_malloc(sizeof(*u_buffer) * (u_len + 1));
+  }
+
+  // Decode utf8 string into Unicode codepoints and convert to lowercase
+  uint32_t codepoint;
+  unsigned i = 0;
+  for (ssize_t j = 0; j < u_len; j++) {
+    // Read unicode codepoint from utf8 string
+    encoded_char = nu_utf8_read(encoded_char, &codepoint);
+    // Transform unicode codepoint to lowercase
+    const char *map = nu_tolower(codepoint);
+
+    // Read the transformed codepoint and store it in the unicode buffer
+    if (map != 0) {
+      uint32_t mu;
+      while (1) {
+        map = nu_casemap_read(map, &mu);
+        if (mu == 0) {
+          break;
+        }
+        u_buffer[i] = mu;
+        ++i;
+      }
+    }
+    else {
+      // If no transformation is needed, just copy the unicode codepoint
+      u_buffer[i] = codepoint;
+      ++i;
+    }
+  }
+
+  // Encode Unicode codepoints back to utf8 string
+  ssize_t reencoded_len = nu_bytenlen(u_buffer, u_len, nu_utf8_write);
+  if (reencoded_len > 0 && reencoded_len <= in_len) {
+    nu_writenstr(u_buffer, u_len, encoded, nu_utf8_write);
+  } else {
+    reencoded_len = 0;
+  }
+
+  // Free heap-allocated memory if needed
+  if (u_buffer != u_stack_buffer) {
+    rm_free(u_buffer);
+  }
+
+  return reencoded_len;
+}
+
+
 // strndup + lowercase in one pass!
 static char *rm_strdupcase(const char *s, size_t len) {
```

Comment:
Seems we should rename this function since it is also unescaping and lowering case

Reply:
Function renamed to `rm_normalize()`